src/pipapo/utils/container.py

```diff
@@ -0,0 +1,250 @@
+
+from __future__ import annotations
+
+from collections.abc import Collection, Iterable, Mapping
+from copy import deepcopy
+from typing import Generic, Self, TypeVar
+
+import numpy as np
+
+from pipapo.utils.iterable_handler import Handlers, ListHandler, NDArrayHandler
+from pipapo.utils.type_hinting import int_np_array
+
+P = TypeVar("P", bound=Mapping)
+D = TypeVar("D", bound=Mapping)
+
+
+HANDLERS = Handlers(
+    {list.__name__: ListHandler(), np.ndarray.__name__: NDArrayHandler()}
+)
+
+
+class Container(Generic[P, D]):
+    """Generic container for properties and data fields."""
+
+    _safe = True
+    _handlers: Handlers = HANDLERS
+
+    def __init__(self, properties: P, data: D | None):
+        """Intiialize container.
+
+        Args:
+            properties: Container properties.
+            data: Container data fields.
+        """
+        self._properties = properties
+        if data is None:
+            self._field_data: D = {}  # type: ignore[assignment]
+        else:
+            self._field_data = data
+
+        self._check_lens()
+
+    def _check_lens(self) -> None:
+        """Check if lengths match."""
+        if self._safe:
+            len(self)
+
+    @property
+    def data(self) -> D:
+        """Get data property."""
+        self._check_lens()
+        return self._field_data
+
+    @data.setter
+    def data(self, new_data: D) -> None:
+        """Set data.
+
+        Args:
+            new_data: New data
+        """
+        self._field_data = new_data
+        self._check_lens()
+
+    def __str__(self) -> str:
+        """Get string representation of the container."""
+        string = type(self).__name__ + f" of size {len(self)}"
+        string += "\n Properties:\n   " + "\n   ".join(self._properties.keys())
+        if self._field_data:
+            string += "\n Data:\n   " + "\n   " + "\n   ".join(self._field_data.keys())
+        return string
+
+    def __len__(self) -> int:
+        """Length of the container."""
+        lengths_property: set = set()
+        lengths_data: set = set()
+
+        for k, v in self._properties.items():
+            try:
+                lengths_property = lengths_property.union((len(v),))
+
+            except TypeError as exc:
+                raise TypeError(
+                    f"Properties '{k}', does not have a `len` method."
+                ) from exc
+
+        for k, v in self._field_data.items():
+            try:
+                lengths_data = lengths_data.union((len(v),))
+
+            except TypeError as exc:
+                raise TypeError(f"Data '{k}', does not have a `len` method.") from exc
+
+        if lengths_data and lengths_data != lengths_property:
+            raise ValueError(
+                f"Data and Properties have different lengths: \n"
+                f" Properties: {lengths_property}\n Data: {lengths_data}"
+            )
+
+        if not lengths_property:
+            return 0
+
+        if len(set(lengths_property)) == 1:
```

Comment:
```suggestion
        if len(lengths_property) == 1:
```
`lengths_property` already is a set, right?